src/StoppingCondition.jl

```diff
@@ -0,0 +1,86 @@
+export AddStoppingCondition
+
+
+function AddStoppingCondition(p::DSProblem, c::T) where T <: AbstractStoppingCondition
+    push!(p.stoppingconditions, c)
+end
+
+CheckStoppingConditions(p::DSProblem) = CheckStoppingConditions(p, p.stoppingconditions)
+function CheckStoppingConditions(p::DSProblem, c::Vector{T}) where T <: AbstractStoppingCondition
+    for condition in c
+        #Stopping conditions return true if optimisation should continue,
+        #false if the condition is met and optimisation should stop
+        if CheckStoppingCondition(p, condition) == false
+            SetStatus(p, condition)
+            return false
+        end
+    end
+
+    return true
+end
+
+function SetStatus(p, s::T) where T <: AbstractStoppingCondition
+    p.status.optimization_status = StoppingConditionStatus(s)
+end
+
+StoppingConditionStatus(::T) where T <: AbstractStoppingCondition = "Unknown stopping condition status"
+
+function SetupStoppingConditions(p::DSProblem)
+    for c in p.stoppingconditions
+        SetupStoppingCondition(p, c)
+    end
+end
+
+SetupStoppingCondition(p::DSProblem, ::AbstractStoppingCondition) = nothing
+
+#===== Built-in stopping conditions =====#
+
+#Iteration limit
+mutable struct IterationStoppingCondition <: AbstractStoppingCondition
+    limit::Int64
+end
+
+StoppingConditionStatus(::IterationStoppingCondition) = "Iteration limit"
+
+CheckStoppingCondition(p::DSProblem, s::IterationStoppingCondition) = p.status.iteration < s.limit
+
+function SetupStoppingCondition(p::DSProblem, s::IterationStoppingCondition)
+    if s.limit == -1 
+        error("Please set a maximum number of iterations")
+    end
+end
+
+"""
+    SetIterationLimit(p::DSProblem, i::Int)
+
+Set the maximum number of iterations to `i`.
+"""
+function SetIterationLimit(p::DSProblem, i::Int)
+    if i < p.status.iteration
+        error("Cannot set iteration limit to lower than the number of iterations that have run")
+    else
+        p.stoppingconditions[1].limit = i
```

Comment:
We should not be assuming an iteration stopping condition is the first condition in the vector. Someone can make the `DSProblem` on their own and put any condition they want as number one, so we should be looping over the vector until we find the iteration limit one and iterate that one (and throw an error if it can't be found).

Reply:
Resolved in 1f0151f, a generic function can be used to return the indexes of all stopping conditions of a certain type.